Create risyou_hanntei.ino

shibata-tomoyo opened the pull request.

shibata-tomoyo committed:
Create risyou_hanntei.ino

4456-kent commented:
タイマー割込みの実装しましょう．delay()関数というのはソフトウェア上で実装する時間制御で，実際の時間はこの関数自体の実行時間も合わせるので設定した時間とのずれがあります．(数マイクロ秒程度ですが，塵つも理論です)実行タイミングを正確に制御するにはタイマー割込みを実装する必要があります．こちらはハードウェアであるマイコンボード上のクロックによって正確な時間を測定できます．またdelay()関数は実行時に何もできないのに対して，タイマー割込みならメインの処理ど同時並行で処理を行うところも特徴です．

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -0,0 +1,171 @@
+
+  Wire.begin(21, 22);
+  Serial.begin(115200);
+
```
mpu6050の測定レンジの設定処理がない．
アドレスは加速度が0x1Cで角速度が0x1B．

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -0,0 +1,171 @@
+  SWAP (accel_t_gyro.reg.y_gyro_h, accel_t_gyro.reg.y_gyro_l);
+  SWAP (accel_t_gyro.reg.z_gyro_h, accel_t_gyro.reg.z_gyro_l);
+
+  float ax = accel_t_gyro.value.x_accel / 16384.0; //FS_SEL_0 16,384 LSB / g
```
生データを割ってる数字から見て測定レンジが±2[G]かな？
合成加速度2.5[G]以上だから±4[G]以上(ていうか±8[G])で設定してほしい．
測定レンジの設定ミスは離床判定できない原因の一つです．(過去に実際ありました)

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -0,0 +1,171 @@
+  Serial.print(az, 2);
+  Serial.print("\t");
+  
+  float gx = accel_t_gyro.value.x_gyro / 131.0;//FS_SEL_0 131 LSB / (°/s)
```
角速度の測定レンジは±2000[deg/s]に設定してほしい．

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -0,0 +1,171 @@
+
+void loop() {
+  //離床判定
+  int an = -1; //移動平均カウントのための変数
```
なんで-1？

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -0,0 +1,171 @@
+  float asqrt_five[5];
+  asqrt_all[an] = sqrt(pow(accel_t_gyro.value.x_accel, 2)+pow(accel_t_gyro.value.y_accel, 2)+pow(accel_t_gyro.value.z_accel, 2)) / 16384.0; //３軸合成加速度
+  for(ai = 0; ai < 5; ai++){
+    asqrt_five[ai] = asqrt_all[an-ai];
```
5個の配列に直で値ぶち込んでいいのでは？

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -0,0 +1,171 @@
+  } else{
+    acnt = 0;
+  }          
+  if(acnt == 5){
```
==で条件判定しないほうがいい．もし何かのバグで5を飛ばして変数が大きくなった場合でも条件判定できるように不等号使ったほうがいい

shibata-tomoyo committed:
Update risyou_hanntei.ino

質問です．このプログラムでは加速度とジャイロを取得していましたが，ジャイロは判定には使用しません．よってジャイロの取得はロガー電装にお任せして，メイン電装では取得する必要がないと感じたのですが，どうでしょうか...？一度ジャイロ関係の処理をすべて消してみましたが，必要あればまた戻します．

shibata-tomoyo committed:
Update risyou_hanntei.ino

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -105,56 +96,35 @@ void setup() {
-  }
-  float aave = asum / 5;
+  if(ai >= 4){
+  float asqrt[ai+1];
```
この置き方すると，変数aiが大きくなればなるほど用意する配列asqrt[]のメモリ占有率が大きくなるから辞めといたほうがいい．

4456-kent replied:
普通に変数aiは0で初期化，4になったら0に戻すっていう風でいい気がする．

4456-kent commented:
> 質問です．このプログラムでは加速度とジャイロを取得していましたが，ジャイロは判定には使用しません．よってジャイロの取得はロガー電装にお任せして，メイン電装では取得する必要がないと感じたのですが，どうでしょうか...？一度ジャイロ関係の処理をすべて消してみましたが，必要あればまた戻します．

OKです．

4456-kent commented:
レンジの設定が変わってないので反映お願いします．あとタイマー割込みの実装の仕方をもう一度言うと，
まず最初にタイマー割込みで使う変数を以下のように用意．
```
volatile int timeCounter1;
hw_timer_t *timer1 = NULL;
portMUX_TYPE timerMux = portMUX_INITIALIZER_UNLOCKED;
```
以下のような関数を定義．
```
void IRAM_ATTR onTimer1(){
  portENTER_CRITICAL_ISR(&timerMux);
  timeCounter1++;
  portEXIT_CRITICAL_ISR(&timerMux);
}
```

次にsetup()関数内で以下のようにタイマー割込みの初期化をする．

```
timer1 = timerBegin(0, 80, true);
timerAttachInterrupt(timer1, &onTimer1, true);
timerAlarmWrite(timer1, 1.0E6 / SAMPLING_RATE, true);
timerAlarmEnable(timer1);  
```

SAMPLING_RATEのところはサンプリングレートを入れた変数を入れたらいい．

次にloop()関数にタイマー割込み処理の呼び出し及びその処理を書く．
```
loop(){
  if(timeCounter1 > 0){
        portENTER_CRITICAL(&timerMux);
        timeCounter1--;
        portEXIT_CRITICAL(&timerMux);
        /*これ以下に割込みで処理するコードを書く*/
  }
}
```

これでやってみて．

shibata-tomoyo committed:
Update risyou_hanntei.ino

shibata-tomoyo commented:
質問です．タイマ割込みについて詳しくありがとうございます．タイマ割込みでの処理は，加速度の値の取得から離床判定までの一連で合っていますか？(というより変数や配列の関係で一連にしないと上手くコンパイルできなかったので...）

4456-kent commented:
値取得と条件判定するのでどれぐらいの処理時間かかるかによる．1/(サンプリングレート)[s]以上かかる処理を割込みに入れることはできないからそれを越えなければいい．(たぶん大丈夫)
あと変数はグローバル変数にすればコンパイルできると思うけど．

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -10,6 +10,7 @@
 
 #define A_FLIGHT 2.50 //しきい値
 boolean BURNING = false; //離床判定によってフェーズがBURNINGに移行．フラグにしてます...
+int SAMPLING_RATE = 400000; //MPU6050のサンプリングレートは400kHz
```
200Hzです．

4456-kent replied:
定数は`const`つけたほうがいいよ．

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -109,9 +139,9 @@ void loop() {
-  float ax = accel.value.x_accel / 16384.0; //FS_SEL_0 16,384 LSB / g
-  float ay = accel.value.y_accel / 16384.0;
-  float az = accel.value.z_accel / 16384.0;
+  float ax = accel.value.x_accel / 4096.0; //FS_SEL_2 4,096 LSB / g
```
こういう数字(4096.0)は変数で置いたほうがいい．もし直前になってレンジを変更したいってなった時にいちいち数字を変えないといけないし，普通に場所によって数字の書き間違いとかよくする(そしてとんでもないことになる)から．

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -121,23 +151,23 @@ void loop() {
-  if(ai >= 4){
-  float aave = (asqrt[ai]+asqrt[ai-1]+asqrt[ai-2]+asqrt[ai-3]+asqrt[ai-4]) / 5;
+  float asqrt[6];
+      for(ai = 0; ai < 5; ai++){
```
124行目でaiに0を代入してるのにここでも0入れてるから重複しとる．

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -121,23 +151,23 @@ void loop() {
-  float aave = (asqrt[ai]+asqrt[ai-1]+asqrt[ai-2]+asqrt[ai-3]+asqrt[ai-4]) / 5;
+  float asqrt[6];
+      for(ai = 0; ai < 5; ai++){
+    asqrt[ai+1] = asqrt[ai];
```
これ空っぽの配列をただ要素番号1だけずらしとるんやけど…

4456-kent commented on risyou_hanntei.ino:
```diff
@@ -121,23 +151,23 @@ void loop() {
+    asqrt[ai+1] = asqrt[ai];
+  }
+  asqrt[0] = sqrt(pow(accel.value.x_accel, 2)+pow(accel.value.y_accel, 2)+pow(accel.value.z_accel, 2)) / 4096.0; //３軸合成加速度
+  float aave = (asqrt[0]+asqrt[1]+asqrt[2]+asqrt[3]+asqrt[4]) / 5;
```
asqrt[0]にしか値入っとらんぞ

shibata-tomoyo committed:
Update risyou_hanntei.ino

shibata-tomoyo committed:
Update risyou_hanntei.ino

shibata-tomoyo committed:
Create risyou_hanntei.ino

shibata-tomoyo committed:
Update risyou_hanntei.ino

shibata-tomoyo committed:
Update risyou_hanntei.ino

ddd3h reviewed (changes requested):
上の方において！
float asqrt[5];
float _asqrt[5];

---------
asqrt[0] = sqrt(pow(accel_t_gyro.value.x_accel, 2)+pow(accel_t_gyro.value.y_accel, 2)+pow(accel_t_gyro.value.z_accel, 2)) / arange;

float aave = (asqrt[0]+asqrt[1]+asqrt[2]+asqrt[3]+asqrt[4]) / 5;

for(ai = 0; ai < 5; ai++){
_asqrt[ai] = asqrt[ai] ;
}
for(ai = 1; ai < 5; ai++){
sqrt[ai] = _sqrt[ai-1] ;
}

shibata-tomoyo committed:
Update risyou_hanntei.ino

shibata-tomoyo committed:
Update risyou_hanntei.ino

shibata-tomoyo committed:
Update risyou_hanntei.ino
